src/python/pants/backend/native/targets/native_library.py

```diff
@@ -0,0 +1,79 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+from pants.backend.native.targets.native_artifact import NativeArtifact
+from pants.base.exceptions import TargetDefinitionException
+from pants.base.payload import Payload
+from pants.base.payload_field import PrimitiveField
+from pants.build_graph.target import Target
+from pants.util.meta import AbstractClass
+
+
+class NativeLibrary(Target, AbstractClass):
+  """A class wrapping targets containing sources for C-family languages and related code."""
+
+  @classmethod
+  def produces_ctypes_dylib(cls, target):
+    return isinstance(target, cls) and bool(target.ctypes_dylib)
+
+  def __init__(self, address, payload=None, sources=None, ctypes_dylib=None,
+               strict_deps=None, fatal_warnings=None, **kwargs):
+
+    if not payload:
+      payload = Payload()
+    sources_field = self.create_sources_field(sources, address.spec_path, key_arg='sources')
+    payload.add_fields({
+      'sources': sources_field,
+      'ctypes_dylib': ctypes_dylib,
```

Comment:
This naming usage of `dylib` is slightly confusing as it clashes with the osx-specific file extension. Is this osx specific at all? Would it improve readability to something like `ctypes_dynamic_library` or the like?

Reply:
*[see another comment about naming here](https://github.com/pantsbuild/pants/pull/5815#discussion_r194590807)*

This isn't osx-specific, it's supposed to state "we can export the contents of this library into an implementation-defined format (e.g. static *or* dynamic lib) with a native ABI to another language". The use of `ctypes_*` everywhere is so it's clear what the capabilities are right now, and is basically an explicit choice to *not* make a final decision on what to call things right now. I especially don't want to adopt names that imply things like static or dynamic libraries at the target level.

I agree that for those reasons and for your reasons, the name `ctypes_dylib` is confusing. Do you think `ctypes_native_library=` might make sense?